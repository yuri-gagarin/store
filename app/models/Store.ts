import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;