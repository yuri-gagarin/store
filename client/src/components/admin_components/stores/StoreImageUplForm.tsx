import React, { useState } from "react";
import { Button, Form } from "semantic-ui-react";
// css imports //
import "./css/storeImgUploadForm.css";

const StoreImageUplForm: React.FC<{}> = (props): JSX.Element => {
  const [ file, setFile ] = useState<File>();

  const handleButtonClick = () => {

  };
  const uploadFile = () => {

  }
  const cancelFile = () => {
    setFile(undefined);
  }

  const fileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files) {
      setFile(e.target.files[0])
    }
  };

  return (
    <div id="storeImgUplFormHolder">
      <Form id="storeImgUploadForm">
        {
          !file ?
          <div id="storeImgInputControlls"> 
            <Button
              as="label"
              content="Choose Image"
              labelPosition="left"
              htmlFor="storeImgFile"
              icon="file"
              onClick={handleButtonClick}
            />
            <input type="file" id="storeImgFile" hidden onChange={fileChange} />
          </div>
          : null
        }
        
        {
          file ?  
          <div id="storeImgUploadControlls">
            <Button 
              id="storeImgCancelBtn"
              as="label"
              content="Cancel"
              labelPosition="left"
              icon="cancel"
              onClick={cancelFile}
            />
            <Button 
              id="storeImgUploadBtn"
              as="label"
              content="Upload"
              icon="upload"
              onClick={uploadFile}
            />
          </div>
          : null
        }
       

       <span>{file ? file.name : "No File"}</span>
      </Form>
    </div>
  );
};

export default StoreImageUplForm;

