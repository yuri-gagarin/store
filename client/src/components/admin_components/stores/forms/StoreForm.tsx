import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";

type Props = {
  title: string;
  description: string;
  handleCreateStore(title: string, description: string): void;
  handleUpdateStore(title: string, description: string): void;
}
type FormState = {
  title: string;
  description: string;
}

const StoreForm: React.FC<Props> = ({ title, description, handleCreateStore, handleUpdateStore }): JSX.Element => {

  const [ newForm, setNewForm ] = useState<boolean>(true)
  const [ formState, setFormState ] = useState<FormState>({ title, description });
  const storeFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  
  const handleTitleChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      title: e.target.value
    });
  };
  const hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateStore(formState.title, formState.description);
    } else {
      handleUpdateStore(formState.title, formState.description);
    }
  };

  useEffect(() => {
    if (title && description) {
      setNewForm(false);
    } else {
      setNewForm(true);
    }
  },  [title, description]);

  useEffect(() => {
    if (storeFormRef.current) {
      const elem = storeFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [storeFormRef]);
  
  return (
    <div id="createStoreFormHolder" ref={storeFormRef}>
      <Form id="createStoreForm">
        <Form.Field>
          <label>Store title</label>
          <input 
            onChange={handleTitleChange} 
            placeholder="Store title here ..." 
            value={formState.title}
          />
        </Form.Field>
        <Form.Field
          id='form-textarea-control-opinion'
          control={TextArea}
          label='Store Description'
          onChange={hadnleDescriptionChange}
          placeholder='Store description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button type='submit' onClick={handleSubmit} content= "Create Store" />
            : <Button type='submit' onClick={handleSubmit} content= "Update Store" />
         }  
      </Form>
    </div>
  );
};

export default StoreForm;