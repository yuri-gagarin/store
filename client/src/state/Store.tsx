import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

