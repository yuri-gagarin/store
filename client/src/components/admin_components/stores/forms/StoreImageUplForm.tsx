import React, { useState, useContext } from "react";
import { Button, Form } from "semantic-ui-react";
// css imports //
import "./css/storeImgUploadForm.css";
// actions //
import { uploadStoreImage } from "../actions/APIstoreActions";
import { Store } from "../../../../state/Store";

const StoreImageUplForm: React.FC<{}> = (props): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, currentStoreData } = state.storeState;
  // local state //
  const [ file, setFile ] = useState<File | null>(null);

  // event handlers and listeners //
  const handleButtonClick = () => {

  };
  const uploadFile = () => {
    if (file) {
      const formData = new FormData();
      const { _id } = currentStoreData;
      formData.append("storeImage", file);
      uploadStoreImage(_id, formData, state, dispatch)
        .then((_) => {
          setFile(null);
        })
        .catch((_) => {
          // log error? show error message ? //
        });
    }
  };
  const cancelFile = () => {
    setFile(null);
  };
  const fileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files) {
      setFile(e.target.files[0])
    }
  };

  return (
    <div id="storeImgUplFormHolder">
      <div><p>Image Uploader</p></div>
      <Form id="storeImgUploadForm">
        {
          !file ?
          <div id="storeImgInputControlls"> 
            <Button
              id="selectStoreImgBtn"
              as="label"
              content="Choose Image"
              labelPosition="left"
              htmlFor="storeImgFile"
              icon="file"
              onClick={handleButtonClick}
            />
            <input type="file" id="storeImgFile" hidden onChange={fileChange} />
          </div>
          : null
        }
        
        {
          file ?  
          <div id="storeImgUploadControlls">
            <Button 
              id="storeImgCancelBtn"
              as="label"
              content="Cancel"
              labelPosition="left"
              icon="cancel"
              onClick={cancelFile}
            />
            <Button 
              id="storeImgUploadBtn"
              as="label"
              content="Upload"
              icon="upload"
              onClick={uploadFile}
              loading={loading}
            />
          </div>
          : null
        }
       

       <span>{file ? file.name : "No File"}</span>
      </Form>
    </div>
  );
};

export default StoreImageUplForm;

