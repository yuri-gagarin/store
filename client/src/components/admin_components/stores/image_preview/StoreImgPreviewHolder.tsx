import React from "react";
// css //
import "./css/storeImgPreviewHolder.css"
// additional components //
import StoreImgPreviewThumb from "./StoreImgPreviewThumb";
// api actions //
import { deleteStoreImage } from "../actions/APIstoreActions";
import { IGlobalAppState, AppAction } from "../../../../state/Store";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>
}
const StoreImgPreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { images } = state.storeState.currentStoreData;

  // delete image handler
  const handleDelete = (_id: string) => {
    deleteStoreImage(_id, state, dispatch);
  };

  return (
    images.length > 0 ?
    <div id="storeImgPreviewHolder">
        <div className="storeImgHolderTitle">Uploaded Images</div>
        {
          images.map((image) => {
            return (
              <StoreImgPreviewThumb 
                key={image._id} 
                _id={image._id} 
                url={image.url} 
                handleDelete={handleDelete}
              />
            );
          })
        }
      </div>
    :
    <div id="storeImgPreviewNone">
      <h3>No images uploaded... you should upload some... </h3>
    </div>
  );
};

export default StoreImgPreviewHolder;