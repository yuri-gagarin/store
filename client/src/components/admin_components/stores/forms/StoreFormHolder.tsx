import React, { useState, useEffect } from "react";
import { Button, Grid } from "semantic-ui-react";
// css imports //
import "./css/storeFormHolder.css";
// additional components //
import StoreForm from "./StoreForm";
import StoreImgPreviewHolder from "../image_preview/StoreImgPreviewThumbs";
import StoreImageUplForm from "./StoreImageUplForm";
// state //
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// api actions //
import { createStore, editStore } from "../actions/APIstoreActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const StoreFormHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const [ formOpen, setFormOpen ] = useState<boolean>(false);
  const [ imgUpload, setImgUpload ] = useState<boolean>(false);
  const [ newForm, setNewForm ] = useState<boolean>(true);

  const { currentStoreData } = state.storeState;
  const { title, description } = currentStoreData;

  const handleCreateStore = (title: string, description: string): void => {
    const storePams = {
      title,
      description,
      storeImages: currentStoreData.images
    };
    createStore(storePams, dispatch)
      .then((success) => {
        if (success) {
          // store created //
          setImgUpload(true);
        } else {
          console.error("error");
        }
      });
  };

  const handleUpdateStore = (title: string, description: string): void => {
    const storeParams = {
      title, description, storeImages: currentStoreData.images
    }
    editStore(currentStoreData._id, storeParams, dispatch, state)
      .then((success) => {
        if (success) {
          setFormOpen(false);
        }
      })
  }

  const handleFormOpen = () => {
    setFormOpen(!formOpen);
  };
  useEffect(() => {
    console.log("form loaded")
  }, []);

  useEffect(() => {
    if (!formOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [formOpen])
  useEffect(() => {
    if (title && description) {
      setNewForm(false);
    }
  }, [title, description]);


  return (
    <div id="storeFormHolder">
      {
        !newForm ?
          <React.Fragment>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="storeFormHolderDetails">
                  <h3>Store title:</h3>
                  <p>{currentStoreData.title}</p>
                </div>
                <div className="storeFormHolderDetails">
                  <h3>Store description:</h3>
                  <p>{currentStoreData.description}</p>
                </div>
                <div className="storeFormHolderTimestamps">
                  <span>Created At: {ConvertDate.international(currentStoreData.createdAt)}</span>
                  <span>Edited At: {ConvertDate.international(currentStoreData.editedAt)}</span>
                </div>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <StoreImgPreviewHolder state={state} dispatch={dispatch} />
              </Grid.Column>
            </Grid.Row>
            <StoreImageUplForm />
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  id="storeFormToggleBtn" onClick={handleFormOpen} content={ !formOpen ? "Open Form" : "Close Form"}></Button>
          {
            formOpen ? <StoreForm 
                        title={title} 
                        description={description} 
                        handleCreateStore={handleCreateStore}
                        handleUpdateStore={handleUpdateStore}
                      /> 
                      : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
}

export default StoreFormHolder;

