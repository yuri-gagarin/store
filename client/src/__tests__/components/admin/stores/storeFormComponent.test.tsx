import React, { Component } from "react";
import { mount, ReactWrapper } from "enzyme";
// component imports //
import StoreFormHolder from "../../../../components/admin_components/stores/forms/StoreFormHolder";
import StoreImageUplForm from "../../../../components/admin_components/stores/forms/StoreImageUplForm";
import StoreImgPreviewHolder, { StoreImgPreviewThumb } from "../../../../components/admin_components/stores/image_preview/StoreImgPreviewThumbs";
// helpers //
import { generateCleanState } from "../../../../test_helpers/miscHelpers";
import { setMockStoreState } from "../../../../test_helpers/storeHelpers";
import StoreForm from "../../../../components/admin_components/stores/forms/StoreForm";
import { Button } from "semantic-ui-react";
import { IGlobalAppState } from "../../../../state/Store";

describe("StoreFormHolder Component tests", () => {
  let wrapper: ReactWrapper; 
  
  describe("Default Form Holder state",  () => {

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount<{}, typeof StoreFormHolder>(<StoreFormHolder state={generateCleanState()} dispatch={jest.fn<React.Dispatch<StoreAction>, []>()}/>)
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.html()).toBeDefined();
    });
    it("Form Should be closed by default", () => {
      const form = wrapper.find(StoreForm);
      expect(form.length).toEqual(0);
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.find(Button);
      expect(toggleButton.length).toEqual(1);
    })
  });

  describe("Form Holder state OPEN - NO Current Store Data",  () => {

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount<{}, typeof StoreFormHolder>(<StoreFormHolder state={generateCleanState()} dispatch={jest.fn<React.Dispatch<StoreAction>, []>()}/>)
    });

    it("Should Properly Mount Form Holder", () => {
      const toggleButton = wrapper.find("#storeFormToggleBtn");
      toggleButton.at(0).simulate("click")
      // open button clicked //
      expect(wrapper.html()).toBeDefined();

    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#storeFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have a Form Create Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreFormCreate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered after toggle button", () => {
      const form = wrapper.find(StoreForm);
      expect(form.length).toEqual(1);
    });
    it("Should NOT have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(StoreImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(0);
    });
    it("Should NOT have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(StoreImageUplForm);
      expect(imgUploadForm.length).toEqual(0);
    });

  });

  describe("Form Holder state OPEN - WITH Current Store Data - NO IMAGES",  () => {
    let state: IGlobalAppState;
    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = setMockStoreState({ currentStore: true });
      wrapper = mount<{}, typeof StoreFormHolder>(<StoreFormHolder state={state} dispatch={jest.fn<React.Dispatch<StoreAction>, []>()}/>)
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.html()).toBeDefined();
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#storeFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered", () => {
      const form = wrapper.find(StoreForm);
      expect(form.length).toEqual(1);
    });
    it("Should have a Form Update Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreFormUpdate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(StoreImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should NOT render any preview images", () => {
      const previewThumb = wrapper.find(StoreImgPreviewThumb);
      expect(previewThumb.length).toEqual(0);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(StoreImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });

  });

  describe("Form Holder state OPEN - WITH Current Store Data - WITH IMAGES",  () => {
    let state: IGlobalAppState;
    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = setMockStoreState({ currentStore: true, storeImages: 3 });
      wrapper = mount<{}, typeof StoreFormHolder>(<StoreFormHolder state={state} dispatch={jest.fn<React.Dispatch<StoreAction>, []>()}/>)
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.html()).toBeDefined();
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#storeFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered", () => {
      const form = wrapper.find(StoreForm);
      expect(form.length).toEqual(1);
    });
    it("Should have a Form Update Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreFormUpdate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(StoreImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should render a correct number of preview images", () => {
      const previewThumb = wrapper.find(StoreImgPreviewThumb);
      const numberOfImages = state.storeState.currentStoreData.images.length;
      expect(previewThumb.length).toEqual(numberOfImages);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(StoreImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });
  });
  
});