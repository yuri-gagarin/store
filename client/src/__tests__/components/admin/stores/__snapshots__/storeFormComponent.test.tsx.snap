// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreFormHolder Component tests Default Form Holder state Should Properly Mount Form Holder 1`] = `
<StoreFormHolder
  dispatch={[MockFunction]}
  state={
    Object {
      "bonusVideoState": Object {
        "currentBonusVideoData": Object {
          "_id": "",
          "createdAt": "",
          "description": "",
          "vimeoURL": "",
          "youTubeURL": "",
        },
        "error": null,
        "loadedBonusVideos": Array [],
        "loading": false,
        "responseMsg": "",
      },
      "productState": Object {
        "currentProductData": Object {
          "_id": "",
          "createdAt": "",
          "description": "",
          "details": "",
          "images": Array [],
          "name": "",
          "price": "",
        },
        "error": null,
        "loadedProducts": Array [],
        "loading": false,
        "responseMsg": "",
      },
      "serviceState": Object {
        "currentServiceData": Object {
          "_id": "",
          "createdAt": "",
          "description": "",
          "images": Array [],
          "name": "",
          "price": "",
        },
        "error": null,
        "loadedServices": Array [],
        "loading": false,
        "responseMsg": "",
      },
      "storeItemState": Object {
        "currentStoreItemData": Object {
          "_id": "",
          "categories": Array [],
          "createdAt": "",
          "description": "",
          "details": "",
          "images": Array [],
          "name": "",
          "price": "",
          "storeId": "",
          "storeName": "",
        },
        "error": null,
        "loadedStoreItems": Array [],
        "loading": false,
        "numberOfItems": 0,
        "responseMsg": "",
      },
      "storeState": Object {
        "currentStoreData": Object {
          "_id": "",
          "createdAt": "",
          "description": "",
          "images": Array [],
          "title": "",
        },
        "error": null,
        "loadedStores": Array [],
        "loading": false,
        "responseMsg": "",
      },
    }
  }
>
  <div
    id="storeFormHolder"
  >
    <GridRow>
      <div
        className="row"
      >
        <GridColumn
          computer={14}
          mobile={16}
          tablet={15}
        >
          <div
            className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
          >
            <Button
              as="button"
              content="Open Form"
              id="storeFormToggleBtn"
              onClick={[Function]}
            >
              <Ref
                innerRef={
                  Object {
                    "current": <button
                      class="ui button"
                      id="storeFormToggleBtn"
                    >
                      Open Form
                    </button>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <button
                        class="ui button"
                        id="storeFormToggleBtn"
                      >
                        Open Form
                      </button>,
                    }
                  }
                >
                  <button
                    className="ui button"
                    id="storeFormToggleBtn"
                    onClick={[Function]}
                  >
                    Open Form
                  </button>
                </RefFindNode>
              </Ref>
            </Button>
          </div>
        </GridColumn>
      </div>
    </GridRow>
  </div>
</StoreFormHolder>
`;
